src/TW/CodeGen/Flow.hs

```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE OverloadedStrings #-}
+module TW.CodeGen.Flow
+    ( makeFileName, makeModule
+    , libraryInfo
+    )
+where
+
+import Prelude
+
+import TW.Ast
+import TW.BuiltIn
+import TW.JsonRepr
+
+import Data.Monoid
+import System.FilePath
+import qualified Data.List as L
+import qualified Data.Text as T
+
+makeFileName :: ModuleName -> FilePath
+makeFileName (ModuleName parts) =
+    (L.foldl' (</>) "" $ map T.unpack parts) ++ ".js"
+
+libraryInfo :: LibraryInfo
+libraryInfo = LibraryInfo "Flow" "flow-typed-wire-utils" "0.1.0.0"
+
+makeModule :: Module -> T.Text
+makeModule m =
+    T.unlines
+    [ "/* @flow */"
+    , "// This file was auto generated by typed-wire. Do not modify by hand"
+    , "// Module Name: " <> printModuleName (m_name m) <> ""
+    , ""
+    , T.intercalate "\n" (map makeImport $ m_imports m)
+    , ""
+    , T.intercalate "\n" (map makeTypeDef $ m_typeDefs m)
+    , ""
+    ]
+
+makeImport :: ModuleName -> T.Text
+makeImport m =
+    "import " <> printModuleName m <> ";"
+
+makeTypeDef :: TypeDef -> T.Text
+makeTypeDef td =
+    case td of
+      TypeDefEnum ed ->
+          makeEnumDef ed
+      TypeDefStruct sd ->
+          makeStructDef sd
+
+makeStructDef :: StructDef -> T.Text
+makeStructDef sd =
+    T.unlines
+    [ "export type " <> fullType <> " ="
+    , "   {| " <> T.intercalate "\n   , " (map makeStructField $ sd_fields sd)
+    , "   |}"
+    , ""
+    ]
+    where
+      fullType =
+          let types = T.intercalate ", " (map unTypeVar $ sd_args sd)
+          in unTypeName (sd_name sd) <> (if length (sd_args sd) > 0 then ("<" <> types <> ">") else types)
+
+makeStructField :: StructField -> T.Text
+makeStructField sf =
+    (unFieldName $ sf_name sf) <> " : " <> (makeType $ sf_type sf)
+
+makeEnumDef :: EnumDef -> T.Text
+makeEnumDef ed =
+    T.unlines
+    [ "export type " <> fullType
+    , "   = " <> T.intercalate "\n   | " (map makeEnumChoice $ ed_choices ed)
+    , ""
+    ]
+    where
+      fullType =
+          let types = T.intercalate ", " (map unTypeVar $ ed_args ed)
+          in unTypeName (ed_name ed) <> (if length (ed_args ed) > 0 then ("<" <> types <> ">") else types)
```

Comment:
This looks like code duplication from above. Can you extract that into a function?